nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,941 @@
+                    Err(TryRecvError::Disconnected) => {
+                        let err = "artifact replication receiver disconnected";
+                        error!(&opctx.log, "{err}");
+                        return json!({"error": err});
+                    }
+                }
+            }
+
+            let sleds = match self.list_sleds(opctx).await {
+                Ok(sleds) => sleds,
+                Err(err) => return json!({"error": format!("{err:#}")}),
+            };
+            let mut counters = TufArtifactReplicationCounters::default();
+            let mut inventory =
+                match self.list_artifacts_from_database(opctx).await {
+                    Ok(inventory) => inventory,
+                    Err(err) => return json!({"error": format!("{err:#}")}),
+                };
+            self.list_artifacts_on_sleds(
+                opctx,
+                &sleds,
+                &mut inventory,
+                &mut counters,
+            )
+            .await;
+            self.list_and_clean_up_local_artifacts(&mut inventory);
+
+            let requests =
+                inventory.into_requests(&sleds, &mut rand::thread_rng());
+            let completed = requests
+                .into_stream(&opctx.log)
+                .buffer_unordered(MAX_REQUEST_CONCURRENCY)
+                .collect::<Vec<_>>()
+                .await;
+            self.insert_debug_requests(completed, &mut counters);
+
+            self.lifetime_counters += counters;
+            serde_json::to_value(TufArtifactReplicationStatus {
+                last_run_counters: counters,
+                lifetime_counters: self.lifetime_counters,
+                request_debug_ringbuf: self.request_debug_ringbuf.clone(),
+                local_repos: self.local.len(),
+            })
+            .unwrap()
+        }
+        .boxed()
+    }
+}
+
+impl ArtifactReplication {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        local_rx: mpsc::Receiver<ArtifactsWithPlan>,
+    ) -> ArtifactReplication {
+        ArtifactReplication {
+            datastore,
+            local: Vec::new(),
+            local_rx,
+            request_debug_ringbuf: Arc::new(VecDeque::new()),
+            lifetime_counters: TufArtifactReplicationCounters::default(),
+        }
+    }
+
+    fn insert_debug_requests(
+        &mut self,
+        mut requests: Vec<TufArtifactReplicationRequest>,
+        counters: &mut TufArtifactReplicationCounters,
+    ) {
+        for request in &requests {
+            counters.inc(request);
+        }
+
+        // `Arc::make_mut` will either directly provide a mutable reference
+        // if there are no other references, or clone it if there are. At this
+        // point there should never be any other references; we only clone this
+        // Arc when serializing the ringbuf to a `serde_json::Value`.
+        let ringbuf = Arc::make_mut(&mut self.request_debug_ringbuf);
```

Comment:
Should we use `get_mut` instead?  (I'm just thinking that if this is wrong for whatever reason the behavior of having multiple different sets of counters floating around might make it hard to debug.)

Reply:
I'd rather not have to panic here if the assumption is wrong. If it does clone, the other copies of `ringbuf` are temporary and used during serialization.

The goal here is to prevent needing to clone the ringbuf every time we go to serialize it to JSON. There are other options here:
- Use `get_mut` and clone if it fails (this is basically what `Arc::make_mut` does)
- Use `get_mut` and panic (I'm not sure what else to do here if it returns None)
- Create an enum similar to `std::borrow::Cow` that can accept a slice for serialization, but deserializes to an owned Vec
- Give up and clone it anyway

Reply:
> If it does clone, the other copies of ringbuf are temporary and used during serialization.

If that's true, agreed, that's not really a problem.  It wasn't obvious to me from the `make_mut` docs which copy would be left in `self` but I guess it has to be the new one (which is what we'd want here)?

Another option would be to skip the update and log a message if this happens.  That might be worse, but again, we're talking about a case we believe can never happen.

I don't have a strong feeling about any of it.

Reply:
Hmm, I like the option of skip-and-log more!

Reply:
Yeah, I mean we're deciding what should happen in a can't-happen (we think) situation and they're all not ideal but likely okay.

At some point when we have built out more of FMA, I'm hoping we can have something sort of like panic in Nexus but where the behavior is just to raise a big red "this shouldn't have been possible" flag (but otherwise proceed).  Sort of like an error-level log message today, but one that's impossible for us to miss in dev/test/support.  There have been other situations like this one and it'd be nice to have the thing treat it as if it really won't happen but in a way that lets us know if we were wrong.